### PR TITLE
feat: buildargs

### PR DESCRIPTION
**Pull request by lorenzomigliorero**
Fix: https://github.com/Dokploy/dokploy/issues/241

### UI

- Create a reusable Secret component
- Refactor the visible/not-visible logic with CSS, more readable and performant
- Apply Secret component to both env and buildArgs fields

image

Docker build envs:

```diff
NODE_VERSION=12.22.12
HOSTNAME=c308d43852b1
YARN_VERSION=1.22.18
SHLVL=1
HOME=/root
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
++BAR=200
++FOO=100
PWD=/
```

Container envs
```diff
++LOREM=300
NODE_VERSION=12.22.12
HOSTNAME=74ec9cb838cf
YARN_VERSION=1.22.18
SHLVL=1
HOME=/root
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
PWD=/
```

### DB

- Add buildArgs field
- Create migration script

### Docker

- Add buildargs param
- Remove createEnvFile util, since envs are already added to the service [here](https://github.com/lorenzomigliorero/dokploy/blob/canary/server/utils/builders/index.ts#L110)

### CODE REVIEW

File: components/dashboard/application/environment/show.tsx [hunk 1]
```diff
@@ -1,42 +1,30 @@
...
 interface Props {
 	applicationId: string;
 }
 
 export const ShowEnvironment = ({ applicationId }: Props) => {
-	const [isEnvVisible, setIsEnvVisible] = useState(true);
-	const { mutateAsync, isLoading } =
-		api.application.saveEnvironment.useMutation();
+	const saveEnvironmentMutation = api.application.saveEnvironment.useMutation();
+	const saveBuildArgsMutation = api.application.saveBuildArgs.useMutation();
```
Comment (Siumauricio) [inline components/dashboard/application/environment/show.tsx]: Instead of multiple routers, use `update` so we can use the same endpoint for both operations

File: components/dashboard/application/environment/show.tsx [hunk 2]
```diff
@@ -1,42 +1,30 @@
...
 const addEnvironmentSchema = z.object({
-	environment: z.string(),
+	env: z.string(),
 });
 
 type EnvironmentSchema = z.infer<typeof addEnvironmentSchema>;
 
+const addBuildArgsSchema = z.object({
+	buildArgs: z.string(),
+});
```
Comment (Siumauricio) [inline components/dashboard/application/environment/show.tsx]: We can remove this schema and add to the main schema and make optional so it will become

```ts
const addEnvironmentSchema = z.object({
	env: z.string(),
	buildArgs: z.string(),
});
```

File: components/dashboard/application/environment/show.tsx [hunk 3]
```diff
@@ -74,94 +63,61 @@ export const ShowEnvironment = ({ applicationId }: Props) => {
...
+										here
+									</a>
+									.
+								</span>
+							}
+							placeholder="NPM_TOKEN=xyz"
+						/>
+					</form>
+				</Form>
+			)}
```
Comment (Siumauricio) [inline components/dashboard/application/environment/show.tsx]: We can safely revert this part to what we had, and just add a condition for dockerfile and we will just have 1 mutation and 1 form, the same as we had but just adding 1 more <FormField for the buildArgs

File: server/utils/builders/docker-file.ts [hunk 4]
```diff
@@ -1,26 +1,25 @@
...
+	const { appName, buildArgs } = application;
 	const dockerFilePath = getBuildAppDirectory(application);
 	try {
 		const image = `${appName}`;
 		const contextPath =
 			dockerFilePath.substring(0, dockerFilePath.lastIndexOf("/") + 1) || ".";
 		const tarStream = tar.pack(contextPath);
 
-		createEnvFile(dockerFilePath, env);
-
```
Comment (Siumauricio) [inline server/utils/builders/docker-file.ts]: Why did you remove this function? this is required

**Review by Siumauricio — CHANGES_REQUESTED** [review]

File: server/utils/builders/docker-file.ts (hunk 4 above)
Comment (lorenzomigliorero) [reply server/utils/builders/docker-file.ts]: I left a comment + screenshot about it in the description.
This is not needed, since envs are already added to the service [here](https://github.com/lorenzomigliorero/dokploy/blob/canary/server/utils/builders/index.ts#L110)

File: components/dashboard/application/environment/show.tsx (hunk 3 above)
Comment (lorenzomigliorero) [reply components/dashboard/application/environment/show.tsx]: I also considered using one form. The only thing that didn't convince me was that the save button gets pushed outside the viewport, and you have to scroll to see it.

image

Should we try using flex-row instead of flex-column?

File: server/utils/builders/docker-file.ts (hunk 4 above)
Comment (Siumauricio) [reply server/utils/builders/docker-file.ts:21]: @lorenzomigliorero Yes but it's not the same, it's what I explained in the Issue that using an .env file the Dockerfile automatically picks up the environment variables.

in fact you can check how we build the docker image from dokploy and you will see that we use an .env file and it automatically uses it as build and as final image.

the args are used in a Docker statement as `ARG` but when you use the `ENV` statement it will use whatever is in the .env file.

Look this example:
`Dockerfile`
https://github.com/Dokploy/dokploy/blob/canary/Dockerfile

How we use the `.env.production.example` from the repo as a `.env.production` in the CI/CD and that are the env's we are using as a Build and as a Final Image
https://github.com/Dokploy/dokploy/blob/canary/.circleci/config.yml#L32

Comment (Siumauricio) [reply server/utils/builders/docker-file.ts]: @lorenzomigliorero Yes but there are many projects where the most common thing is that when you use a docker image to have an .env file and many times there are validations that if this file does not exist the build is broken, I include myself because I have run other projects where if this file does not exist it throws an exception, example:

https://github.com/NetSocialOSS/Harbinger/blob/main/main.go#L20

Comment (lorenzomigliorero) [reply server/utils/builders/docker-file.ts]: Oh yeah, sure, I'll revert the change.
I thought it was related to the _buildargs_ discussion, as you commented [here](https://github.com/Dokploy/dokploy/issues/241#issuecomment-2241377235).

On it!

File: components/dashboard/application/environment/show.tsx (hunk 3 above)
Comment (lorenzomigliorero) [reply components/dashboard/application/environment/show.tsx]: Done! I don't think having the save btn outside the viewport is a big deal. Let me know!

File: components/dashboard/application/environment/show.tsx (hunk 1 above)
Comment (lorenzomigliorero) [reply components/dashboard/application/environment/show.tsx]: Updated keeping one operation!

**Review by Siumauricio — APPROVED** [review]
